sklearn/tree/_criterion.pyx

```diff
@@ -957,11 +954,108 @@ cdef class MSE(RegressionCriterion):
 
         for k in range(self.n_outputs):
             impurity_left[0] -= (sum_left[k] / self.weighted_n_left) ** 2.0
-            impurity_right[0] -= (sum_right[k] / self.weighted_n_right) ** 2.0 
+            impurity_right[0] -= (sum_right[k] / self.weighted_n_right) ** 2.0
 
         impurity_left[0] /= self.n_outputs
         impurity_right[0] /= self.n_outputs
 
+cdef class MAE(RegressionCriterion):
+    """Mean absolute error impurity criterion
+    """
+    cdef void node_value(self, double* dest) nogil:
+        """Computes the node value of samples[start:end] into dest."""
+        cdef SIZE_t start = self.start
+        cdef SIZE_t end = self.end
+        compute_weighted_median(dest, start, end, self.sample_weight, self.y,
+                                self.samples, self.y_stride, self.n_outputs)
+
+    cdef double node_impurity(self) nogil:
+        """Evaluate the impurity of the current node, i.e. the impurity of
+           samples[start:end]"""
+        cdef double* medians = NULL
+        medians = <double*> calloc(self.n_outputs, sizeof(double))
```

Comment:
You could allocate `medians` once at the beginning of the construciton of the tree, and then free it afterwards. This would make it possible to use `safe_realloc` while avoiding the allocation/deallocation dance that you are currently doing.